### PR TITLE
[2023-09-21] Normalization

### LINKED ISSUE #20: [07] Metric, Gradient Descent, Normalization
## Summary
<table>
  <tr>
    <th>Date</th>
    <td>2023-09-19</td>
  </tr>
  <tr>
    <th>Subject</th>
    <td>ML/DL</td>
  </tr>
</table>

## Question
1. 사용한 Metric과 Metric 선정 이유를 설명하시오
2. Gradient Descent란?
3. Normalization의 동작은?

### PR DESCRIPTION
**Pull request by ksumini**
### PR Summary

9월 19일 화요일 진행한 ML-DL 파트의 Normalization에 대한 부분을 정리했습니다.
피드백 부탁드립니다~~!!

### CODE REVIEW

File: 07-ML & DL/Normalization.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+</br>
+
+
+## 📑 꼬리질문
+### 정규화를 통해 값을 Scaling하는 이유에 대해서 설명해주세요.
+
+<p align="center"><img src="https://github.com/RecoRecoNi/Tech-Interview/assets/70088803/2481a07e-7ebf-45e5-9d0b-54c9f14ac35a" height="300" width="600px"></p>
+
+특성들의 스케일이 매우 다르면 오른쪽 그림과 같이 비용 함수 공간의 모양이 길쭉한(elongated) 모양일 수 있습니다.
```
Comment (zsmalla) [inline 07-ML & DL/Normalization.md:33]: 여기에 길쭉한 타원형의 비용 함수 공간인 경우 로스가 수렴하는 데 더 많은 시간이 소요된다는 점도 명시해두면 좋을 것 같아요!

**Review by zsmalla — APPROVED** [review]
수민님 고생하셨습니다! 내용 상으로 문제될 부분도 없어보이고 Latex도 전부 잘 반영된 것 같습니다. 추가되었으면 하는 부분 코멘트 달아두었습니다. 혹시 BatchNorm 공부하시면서 저희 이번이 의문점이 나왔던 부분에 대해서 정리하시게 된다면 해당 내용도 간략하게 추가하면 좋을 것 같습니다. 시간 나실 때 반영 부탁드릴께요!

File: 07-ML & DL/Normalization.md (hunk 1 above)
Comment (ksumini) [reply 07-ML & DL/Normalization.md:33]: 지수님 바로 아래 문장에 있어요!! 혹시 더 추가해야할까요?!

**Review by Woo-Yeol — APPROVED** [review]
저번에 함께 얘기했었던 배치 놈을 진행할 때 사용되는 통계량이 학습시와 인퍼런스시에 어떻게 다른지 내용이 추가되면 좋을 것 같습니다!

### Batch Norm의 학습시와 인퍼런스시의 사용되는 통계량의 차이는 무엇일까요?
학습시에는 배치 단위의 데이터를 기준으로 통계량(평균,분산)을 계산하고 인퍼런스 시에는 학습시 배치마다 얻었던 통계량의 이동 평균을 사용하여 각 평균과 분산을 구한다.

배치의 이동 평균(Moving Average)을 사용하여 해결한다. 즉, inference 전에 학습 시에 미리 미니 배치를 뽑을 때 Sample Mean 및 Sample Variance 를 이용하여 각각의 이동 평균을 구해놔야한다. 위 수식에서 Inference 시에 평균은 각 미니 배치에서 구한 평균들의 평균을 사용하고, 분산은 분산의 평균에 m/(m-1) 을 곱해주게 된다. 이를 곱하는 이유는 통계학적으로 unbiased variance 에는 "Bessel's correction"을 통해 보정을 해주는 것이다. 이는 학습 전체 데이터에 대한 분산이 아니라 미니 배치들의 분산을 통해 전체 분산을 추정할 때 통계학 적으로 보정을 위해 Bessel의 보정 값을 곱해주는 방식으로 추정하기 때문이다...
[티스토리 : Batch Normalization](https://eehoeskrap.tistory.com/430)

**Comment by ksumini** [comment]
> 저번에 함께 얘기했었던 배치 놈을 진행할 때 사용되는 통계량이 학습시와 인퍼런스시에 어떻게 다른지 내용이 추가되면 좋을 것 같습니다!
> 
> ### Batch Norm의 학습시와 인퍼런스시의 사용되는 통계량의 차이는 무엇일까요?
> 학습시에는 배치 단위의 데이터를 기준으로 통계량(평균,분산)을 계산하고 인퍼런스 시에는 학습시 배치마다 얻었던 통계량의 이동 평균을 사용하여 각 평균과 분산을 구한다.
> 
> 배치의 이동 평균(Moving Average)을 사용하여 해결한다. 즉, inference 전에 학습 시에 미리 미니 배치를 뽑을 때 Sample Mean 및 Sample Variance 를 이용하여 각각의 이동 평균을 구해놔야한다. 위 수식에서 Inference 시에 평균은 각 미니 배치에서 구한 평균들의 평균을 사용하고, 분산은 분산의 평균에 m/(m-1) 을 곱해주게 된다. 이를 곱하는 이유는 통계학적으로 unbiased variance 에는 "Bessel's correction"을 통해 보정을 해주는 것이다. 이는 학습 전체 데이터에 대한 분산이 아니라 미니 배치들의 분산을 통해 전체 분산을 추정할 때 통계학 적으로 보정을 위해 Bessel의 보정 값을 곱해주는 방식으로 추정하기 때문이다... [티스토리 : Batch Normalization](https://eehoeskrap.tistory.com/430)

@Woo-Yeol 님, 해당 내용을 좀 더 줄여서 쓴 내용으로 다시 커밋했는데 괜찮을까요?!(확인 부탁드려요!) 찾아보니까 **단순 이동 평균**이 아닌 **지수 이동 평균**을 사용함으로써 전체 데이터 셋을 고려할 수 있기 때문에 **지수 이동 평균**을 사용한다고 하더라구요! 

(**단순 이동 평균**은 이전 N개의 데이터 포인트만을 고려하고, 이전 데이터의 영향을 완전히 잊어버리는 반면, **지수 이동 평균**은 지수적인 감쇠를 통해 이전 데이터의 영향을 조금씩 유지하면서도 최근 데이터에 높은 가중치를 부여합니다. 이로 인해 **지수 이동 평균**은 데이터의 추세를 파악하는 데 더 적합하며, 전체 데이터셋을 고려하여 평균을 계산합니다.)

File: 07-ML & DL/Normalization.md (hunk 1 above)
Comment (zsmalla) [reply 07-ML & DL/Normalization.md:33]: 오 아닙니다! 감사해요~

**Comment by Woo-Yeol** [comment]
> > 저번에 함께 얘기했었던 배치 놈을 진행할 때 사용되는 통계량이 학습시와 인퍼런스시에 어떻게 다른지 내용이 추가되면 좋을 것 같습니다!
> > ### Batch Norm의 학습시와 인퍼런스시의 사용되는 통계량의 차이는 무엇일까요?
> > 학습시에는 배치 단위의 데이터를 기준으로 통계량(평균,분산)을 계산하고 인퍼런스 시에는 학습시 배치마다 얻었던 통계량의 이동 평균을 사용하여 각 평균과 분산을 구한다.
> > 배치의 이동 평균(Moving Average)을 사용하여 해결한다. 즉, inference 전에 학습 시에 미리 미니 배치를 뽑을 때 Sample Mean 및 Sample Variance 를 이용하여 각각의 이동 평균을 구해놔야한다. 위 수식에서 Inference 시에 평균은 각 미니 배치에서 구한 평균들의 평균을 사용하고, 분산은 분산의 평균에 m/(m-1) 을 곱해주게 된다. 이를 곱하는 이유는 통계학적으로 unbiased variance 에는 "Bessel's correction"을 통해 보정을 해주는 것이다. 이는 학습 전체 데이터에 대한 분산이 아니라 미니 배치들의 분산을 통해 전체 분산을 추정할 때 통계학 적으로 보정을 위해 Bessel의 보정 값을 곱해주는 방식으로 추정하기 때문이다... [티스토리 : Batch Normalization](https://eehoeskrap.tistory.com/430)
> 
> @Woo-Yeol 님, 해당 내용을 좀 더 줄여서 쓴 내용으로 다시 커밋했는데 괜찮을까요?!(확인 부탁드려요!) 찾아보니까 **단순 이동 평균**이 아닌 **지수 이동 평균**을 사용함으로써 전체 데이터 셋을 고려할 수 있기 때문에 **지수 이동 평균**을 사용한다고 하더라구요!
> 
> (**단순 이동 평균**은 이전 N개의 데이터 포인트만을 고려하고, 이전 데이터의 영향을 완전히 잊어버리는 반면, **지수 이동 평균**은 지수적인 감쇠를 통해 이전 데이터의 영향을 조금씩 유지하면서도 최근 데이터에 높은 가중치를 부여합니다. 이로 인해 **지수 이동 평균**은 데이터의 추세를 파악하는 데 더 적합하며, 전체 데이터셋을 고려하여 평균을 계산합니다.)

@ksumini 수민님 늦게 확인해서 죄송합니다 ㅠㅠㅠㅠ 확인해보았을 때 너무 정리를 잘해주셔서 너무 도움이 많이 되었습니다. 지수 이동 평균을 이용하는 이유까지 한 마디로 정의해주셔서 더 좋았습니다! 감사합니다 수민님!
